### PR TITLE
fix: use not strictly translation when remote ApisixRoute

### PR DESCRIPTION
**Pull request by gxthrj**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

- Related issues
#540

### CODE REVIEW

File: pkg/ingress/apisix_route.go [hunk 1]
```diff
@@ -139,7 +139,11 @@ func (c *apisixRouteController) sync(ctx context.Context, ev *types.Event) error
 			return err
 		}
 	} else {
-		tctx, err = c.controller.translator.TranslateRouteV2alpha1(ar.V2alpha1())
+		if ev.Type != types.EventDelete {
```
Comment (tokers) [inline pkg/ingress/apisix_route.go:142]: We may description the reason here.

**Comment by tokers** [comment]
@gxthrj Need some test cases to cover it.

File: pkg/kube/translation/apisix_route.go [hunk 2]
```diff
@@ -304,6 +346,24 @@ func (t *translator) translateRouteMatchExprs(nginxVars []configv2alpha1.ApisixR
...
+		sr := apisixv1.NewDefaultStreamRoute()
+		name := apisixv1.ComposeStreamRouteName(ar.Namespace, ar.Name, part.Name)
+		sr.ID = id.GenID(name)
+		sr.ServerPort = part.Match.IngressPort
+		ups, err := t.translateUpstreamNotStrictly(ar.Namespace, backend.ServiceName, backend.Subset, backend.ServicePort.IntVal)
+		if err != nil {
+			return err
+		}
+		sr.Upstream = ups
+		ctx.addStreamRoute(sr)
```
Comment (tokers) [inline pkg/kube/translation/apisix_route.go:362]: Should also add the `ups` to `ctx`.

File: pkg/kube/translation/util.go [hunk 3]
```diff
@@ -110,6 +110,14 @@ loop:
 	return svc.Spec.ClusterIP, svcPort, nil
 }
 
+// translateUpstreamNotStrictly is not strictly translation, only generate ID and Name for delete Event.
```
Comment (tokers) [inline pkg/kube/translation/util.go:113]:
```suggestion
// translateUpstreamNotStrictly translates Upstream nodes with a loose way, only generate ID and Name for delete Event.
```

File: pkg/kube/translation/apisix_route.go (hunk 2 above)
Comment (gxthrj) [reply pkg/kube/translation/apisix_route.go:362]: ref to：https://github.com/apache/apisix-ingress-controller/blob/master/pkg/kube/translation/apisix_route.go#L333
Are you sure need to add `ups` to `ctx`?

Comment (tokers) [reply pkg/kube/translation/apisix_route.go:361]: We should.

By the way, the code in your link should also be changed as now stream_route can use upstream_id.

File: pkg/kube/translation/apisix_route.go [hunk 4]
```diff
@@ -329,8 +389,9 @@ func (t *translator) translateTCPRoute(ctx *TranslateContext, ar *configv2alpha1
 		if err != nil {
 			return err
 		}
-		sr.Upstream = ups
+		sr.UpstreamId = ups.ID
 		ctx.addStreamRoute(sr)
+		ctx.addUpstream(ups)
```
Comment (tokers) [inline pkg/kube/translation/apisix_route.go:394]: Should check whether it existed already.

File: pkg/kube/translation/apisix_route.go (hunk 2 above)
Comment (gxthrj) [reply pkg/kube/translation/apisix_route.go:361]: Fixed.

File: pkg/kube/translation/apisix_route.go (hunk 4 above)
Comment (gxthrj) [reply pkg/kube/translation/apisix_route.go]: Done.

**Review by tokers — APPROVED** [review]

**Review by tokers — CHANGES_REQUESTED** [review]
Need some test cases.

**Review by tokers — APPROVED** [review]

**Review by tao12345666333 — COMMENTED** [review]
It's red, PTAL

**Comment by gxthrj** [comment]
Oh my god, what did I do?

**Comment by gxthrj** [comment]
CI failed caused by APISIX:dev has changed the response of `admin api`, I re-built the dev image, and re-run all of CI.

File: test/e2e/scaffold/k8s.go [hunk 5]
```diff
@@ -29,11 +28,12 @@ import (
 	"github.com/stretchr/testify/assert"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 type counter struct {
-	Count string `json:"count"`
+	Count intstr.IntOrString `json:"count"`
```
Comment (tokers) [inline test/e2e/scaffold/k8s.go:36]: Are you sure the marshaling and unmarshaling can be done?

type IntOrString struct {
	Type   Type   `protobuf:"varint,1,opt,name=type,casttype=Type"`
	IntVal int32  `protobuf:"varint,2,opt,name=intVal"`
	StrVal string `protobuf:"bytes,3,opt,name=strVal"`
}

Comment (gxthrj) [reply test/e2e/scaffold/k8s.go]: This can not be work. 
Done. Customize the `intOrString` type.

**Review by nic-chen — APPROVED** [review]

**Pull request merged**